longformer/longformer.py

```diff
@@ -157,38 +180,90 @@ def forward(
-            if self.attention_mode == 'tvm':
-                d_mask = diagonaled_mm_tvm(ones, float_mask, self.attention_window, self.attention_dilation, False, 0, False)
+            if self.attention_mode == "tvm":
+                d_mask = diagonaled_mm_tvm(
+                    ones, float_mask, self.attention_window, self.attention_dilation, False, 0, False,
+                )
             elif self.attention_mode == "sliding_chunks":
                 d_mask = sliding_chunks_matmul_qk(ones, float_mask, self.attention_window, padding_value=0)
             elif self.attention_mode == "sliding_chunks_no_overlap":
                 d_mask = sliding_chunks_no_overlap_matmul_qk(ones, float_mask, self.attention_window, padding_value=0)
 
             attn_weights += d_mask
         assert list(attn_weights.size())[:3] == [bsz, seq_len, self.num_heads]
-        assert attn_weights.size(dim=3) in [self.attention_window * 2 + 1, self.attention_window * 3]
+        assert attn_weights.size(dim=3) in [
+            self.attention_window * 2 + 1,
+            self.attention_window * 3,
+        ]
 
         # the extra attention
         if extra_attention_mask is not None:
             selected_k = k.new_zeros(bsz, max_num_extra_indices_per_batch, self.num_heads, self.head_dim)
             selected_k[selection_padding_mask_nonzeros] = k[extra_attention_mask_nonzeros]
             # (bsz, seq_len, num_heads, max_num_extra_indices_per_batch)
-            selected_attn_weights = torch.einsum('blhd,bshd->blhs', (q, selected_k))
+            selected_attn_weights = torch.einsum("blhd,bshd->blhs", (q, selected_k))
             selected_attn_weights[selection_padding_mask_zeros[0], :, :, selection_padding_mask_zeros[1]] = -10000
             # concat to attn_weights
-            # (bsz, seq_len, num_heads, extra attention count + 2*window+1)
+            # (bsz, seq_len, num_heads, max_num_extra_indices_per_batch + 2*window+1)
             attn_weights = torch.cat((selected_attn_weights, attn_weights), dim=-1)
+
+        if position_bias is None and self.has_relative_attention_bias:
+            window_relative_position = torch.arange(
+                -self.attention_window, self.attention_window + 1, 1, dtype=torch.long, device=attn_weights.device,
+            )
+            window_position_bias = (
+                self.relative_attention_bias(
+                    relative_position_bucket(window_relative_position, num_buckets=self.relative_attention_num_buckets)
+                )
+                .permute(1, 0)[None, None, :, :]
+                .repeat(bsz, seq_len, 1, 1)
+            )  # (bsz, seq_len, num_heads, 2*window+1)
+            perm_global_position_bias = attn_weights.new_zeros(
+                bsz, max_num_extra_indices_per_batch, seq_len, self.num_heads
+            )  # (bsz, max_num_extra_indices_per_batch, seq_len, num_heads)
+            if extra_attention_mask is not None:
```

Comment:
Global position bias is a bit more complex. We first get the memory position from `extra_attention_mask_nonzeros`, then compute the query position using arrange. Their diff is the relative position. But this "sparse" one vector for each global token in the batch. So we later put it back into the shape of `(bsz, max_num_extra_indices_per_batch, ...)` using the index information from `selection_padding_mask_nonzeros`